ethereum/nft/trades/insert_cryptopunks.sql

```diff
@@ -0,0 +1,165 @@
+CREATE OR REPLACE FUNCTION nft.insert_cryptopunks(start_ts timestamptz, end_ts timestamptz=now(), start_block numeric=0, end_block numeric=9e18) RETURNS integer
+LANGUAGE plpgsql AS $function$
+DECLARE r integer;
+BEGIN
+
+WITH rows AS (
+    INSERT INTO nft.trades (
+	block_time,
+	nft_project_name,
+	nft_token_id,
+	platform,
+	platform_version,
+	category,
+	evt_type,
+	usd_amount,
+	seller,
+	buyer,
+	original_amount,
+	original_amount_raw,
+	original_currency,
+	original_currency_contract,
+	currency_contract,
+	nft_contract_address,
+	exchange_contract_address,
+	tx_hash,
+	block_number,
+	tx_from,
+	tx_to,
+	trace_address,
+	evt_index,
+	trade_id
+    )
+
+    SELECT
+        trades.evt_block_time AS block_time,
+        'CryptoPunks' AS nft_project_name,
+        CAST(trades."punkIndex" AS TEXT) AS nft_token_id,
+        'LarvaLabs Contract' AS platform,
+        '1' AS platform_version,
+        'Buy' AS category,
+        'Trade' AS evt_type,
+        trades.value / 10 ^ 18 * p.price AS usd_amount,
+        trades."fromAddress" AS seller,
+        trades."toAddress" AS buyer,
+        trades.value / 10 ^ 18 AS original_amount,
+        trades.value AS original_amount_raw,
+        'ETH' AS original_currency,
+        '\x0000000000000000000000000000000000000000'::bytea AS original_currency_contract,
+        '\xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2'::bytea AS currency_contract,
+        trades.contract_address AS nft_contract_address,
+        trades.contract_address AS exchange_contract_address,
+        trades.evt_tx_hash AS tx_hash,
+        trades.evt_block_number AS block_number,
+        tx."from" AS tx_from,
+        tx."to" AS tx_to,
+        NULL::integer[] AS trace_address,
+        trades.evt_index,
+        row_number() OVER (PARTITION BY 4, 18, 23, 6) AS trade_id
```

Comment:
Something else just struck me. You need to add an order by clause inside of this partition ([see here](https://github.com/duneanalytics/abstractions/blob/master/ethereum/dex/trades/insert_1inch_lp.sql#L56)). We have had issues where this partition forms a record set of > 1 but the order of the record set was random. That means that if the function was run twice on the same set of data it could insert a given entry multiple times because its `trade_id` is different.